### DISCUSSION TITLE
Web3Modal: React Native + wagmi

### DISCUSSION THREAD

**Discussion by thiagobrez** [WalletConnect/walletconnect-monorepo#3762] [Developer Support]
Hey team! First of all, thanks for the hard work put in here.

Are there any plans for Web3Modal to support a React Native + wagmi config?

I'm trying to build a dApp with WalletConnect, but the only recommended React Native approach in the docs seem to be using WalletConnectModal. It works great on itself, but doesn't seem to be able to plug in into wagmi.

From a react native standpoind, there seems to be 2 possible options:

1. Use WalletConnectModal: a workaround seems to be retrieving `const {provider} = useWalletConnectModal()`, creating a custom connector in wagmi passing the provider, and opening the modal from inside the custom connector. Haven't tested this yet though, would you recommend?
2. Use EthereumProvider: instantiating a new EthereumProvider and passing to a custom connector in wagmi seems like a perfect solution, but this is currently not possible because EthereumProvider relies internally on `@walletconnect/modal` (web version), so it fails on RN because it tries to render HTML. Would be amazing if we had a EthereumProvider made for RN, where it would call the already existing `@walletconnect/modal-react-native` internally. Was this ever discussed?

I would happily send PRs contributing if you could point me in the right direction - thanks!

**Comment by glitch-txs** [comment]
Hi are you already using Wagmi in your app?
We're working on a RN + Viem integration which is a little bit more low level than wagmi.

**Reply by thiagobrez** [reply]
Hi @glitch-txs! I'm not currently working in an app itself, but on educational content for RN devs coming to web3. I appreciate this new integration you're working on and I'm excited to see that, but at the same time, I think it would be beneficial for the community as a whole to have a good way to integrate good tools they're already using in the web when coming to native

**Reply by glitch-txs** [reply]
I'm not very familiar with the RN community, is wagmi used? AFAIK wagmi have a set of connectors that only work with the browser.
But we can request this to our engineer if that's something devs want

**Reply by thiagobrez** [reply]
Not much used yet I'd say exactly for this reason - no built-in connectors that work with RN. Wagmi itself does not depend on browser APIs, so that's basically the only thing keeping devs from using it on RN too. That's why I created this discussion, to try and find a way to get a WalletConnect mobile connector :D

And it seems like the perfect fit because:
- WalletConnect already has the RN modal UI: `@walletconnect/modal-react-native`
- WalletConnect already has the EthereumProvider, which is EIP-1193 compliant, so fits `viem` Transports, and consequently `wagmi` connectors perfectly

The problem IMO seems to be that EthereumProvider depends on the web modal UI. See here: https://github.com/WalletConnect/walletconnect-monorepo/blob/2a4188ee986c3e26a37241601733bcb92018c580/providers/ethereum-provider/src/EthereumProvider.ts#L11C75-L11C75

If we had something like "EthereumProvider for React Native", that depended on the already existing React Native modal UI, instead of the web one, I believe a connector for wagmi could be easily created.

Then, we could provide this new Ethereum Provider for React Native implementation alongside WalletConnect docs, and also try to merge upstream into wagmi.

I also created a discussion in wagmi earlier today to talk about these exact same things, if you want to take a peak: https://github.com/wagmi-dev/wagmi/discussions/3099

What do you think?

**Reply by glitch-txs** [reply]
We already have WalletConnectModal for RN which is basically the same as ethereum-provider but with different UI.
Wagmi uses React-Query is this supported by RN?

**Reply by glitch-txs** [reply]
just checked, tbh I don't think there's much to it, we'll be handling chain and account switching for Web3Modal RN, and you could add [React Query](https://tanstack.com/query/v4/docs/react/react-native) on top of Viem, this would basically be Wagmi and you won't need to install unrelated SDKs/connectors
There might be also room for an extra package maybe here? Something that adds React Query on top of Web3Modal

cc @ignaciosantise

**Reply by thiagobrez** [reply]
>We already have WalletConnectModal for RN which is basically the same as ethereum-provider but with different UI.

Ok, so in that case, I'll give a try to the first approach I posted in the first comment of the discussion: getting the provider from `useWalletConnectModal()` and passing that down to a new wagmi custom connector

>Wagmi uses React-Query is this supported by RN?

Yep!

Really appreciate your time and discussion, will come back with updates soon :)

**Reply by thiagobrez** [reply]
>we'll be handling chain and account switching for Web3Modal RN, and you could add [React Query](https://tanstack.com/query/v4/docs/react/react-native) on top of Viem

Cool, that's great to see and definitely an improvement. But I'd still argue there's reasons to use wagmi:
1. Yes, we can build it with React Query, but it's a lot of rework
2. Teams often have a hybrid setup with React Native + react-native-web, and it would be just a matter of switching connectors, being able to reuse most of the existing code

**Reply by glitch-txs** [reply]
I'd be down to this idea if the connectors where on a different package (some of our concerns [see issue](https://github.com/wagmi-dev/wagmi/issues/3073) ), also Wagmi is very web focused so if they release a version that breaks on RN that wouldn't be beneficial.

I think a small package that wraps Viem function with react query would be cleaner.

  Feel free to open an issue on this as a feature request on the React Native repo, Nacho is currently driving the project :)

**Reply by glitch-txs** [reply]
I've pinged Wagmi team, we'll look into this, thanks for the feedback 🤝

**Reply by thiagobrez** [reply]
>I'd be down to this idea if the connectors where on a different package (some of our concerns https://github.com/wagmi-dev/wagmi/issues/3073 )

Agree!

>I've pinged Wagmi team, we'll look into this, thanks for the feedback 🤝

Ok, I'll wait - thanks for taking the time!
